
syntax = "proto3";

package workerPackage;

service Worker {
    rpc setDriverPort(driverPort) returns (status);
    rpc map(kmeansInput) returns (status);
    rpc reduce (kmeansReduce) returns (status);
    rpc die (empty) returns (status);
    rpc sendPartitionedData(PartitionRequest) returns (MapperDataList);
}

message empty {} // empty message

message kmeansReduce {
    int32 id = 1; // Reduce id
    int32 mapNums = 2;
}

message status {
    int32 code = 1;
    string msg = 2;
}

message kmeansInput {
    string path = 1; // File path
    int32 mapID = 2; // ID of map opperation
    int32 numClusters = 3; //
    repeated float centroids = 4;
    int32 numReducers = 5; //
}

message driverPort {
    int32 port = 1;
}

message PartitionRequest {
    int32 reducerID = 1;
}

message MapperData {
    int32 reducerID = 1;       
    string data = 2; 
}

message MapperDataList {
    repeated string data = 1; // List of strings
}
